metadata-ingestion/src/datahub/api/entities/external/external_entities.py

```diff
@@ -0,0 +1,252 @@
+        If the urn is already in the set, we don't need to add it again, but
+        that is not a conflict.
+        """
+        if urn.urn() in self.urns:
+            return False
+
+        # Detect the family of the urns in the existing set
+        detected_family = None
+        for existing_urn in self.urns:
+            try:
+                parsed_urn = Urn.from_string(existing_urn)
+                family = parsed_urn.entity_type
+                if detected_family is None:
+                    detected_family = family
+                elif detected_family != family:
+                    logger.warning(
+                        f"Detected family {detected_family} is not equals to {family}"
+                    )
+                    return True
+            except ValueError:
+                # Not a valid URN
+                logger.warning(f"Invalid URN {existing_urn} in LinkedResourceSet")
+                return True
+        try:
+            parsed_urn = urn
+            if (
+                detected_family is not None
+                and parsed_urn.entity_type != detected_family
+            ):
+                logger.warning(
+                    f"Detected family {detected_family} is not equals to parsed_urn's family: {parsed_urn.entity_type}"
+                )
+                return True
+        except ValueError:
+            # Not a valid URN
+            logger.warning(f"Invalid URN: {urn} in LinkedResourceSet")
+            return True
+        detected_case_sensitivity = self._detect_case_sensitivity()
+        if (
+            LinkedResourceSet.detect_case_sensitivity(urn) != detected_case_sensitivity
+            and detected_case_sensitivity != CaseSensitivity.MIXED
+        ):
+            logger.warning(
+                f"Detected case sensitivity {detected_case_sensitivity} is not equals to {LinkedResourceSet.detect_case_sensitivity(urn)}"
+            )
+            return True
+
+        return False
+
+    def add(self, urn: Union[str, Urn]) -> bool:
+        """
+        Adds a URN to the set.
+        Returns True if the URN was added, False if it was already in the set.
+        Raises a ValueError if the URN is in conflict with the existing set.
+        """
+        # Deduplicate the URNs if we have somehow duplicate items from concurrent runs
+        self.urns = list(set(self.urns))
+        if isinstance(urn, str):
+            urn = Urn.from_string(urn)
+        if self._has_conflict(urn):
+            raise ValueError(f"Conflict detected when adding URN {urn} to the set")
+        if urn.urn() not in self.urns:
+            self.urns.append(urn.urn())
+            return True
+        return False
+
+
+class ExternalEntity:
+    """
+    An ExternalEntity is a representation of an entity that external to DataHub
+    but could be linked to one or more DataHub entities.
+    """
+
+    @abstractmethod
+    def is_managed_by_datahub(self) -> bool:
```

Comment:
Out of curiosity: what's an example of an external entity being managed by datahub?

Reply:
In DataHub Cloud you can enable automations to sync tags from datahub to external systems like Snowflake, Bigquery and in that case this flag is set to true which means it is controlled by DataHub.

Reply:
then

> An ExternalEntity is a representation of an entity that external to DataHub and could be linked **to zero, one or more** DataHub entities.